### ISSUE TITLE
Does not work with Java 17

### ISSUE THREAD

**Issue by mickle-ak** [easymock/easymock#274]
I clone EasyMock repository to me local and start unit tests with Java 17: failed 126, passed: 562

java.lang.ExceptionInInitializerError
	at org.easymock.internal.ClassProxyFactory.createEnhancer(ClassProxyFactory.java:233)
	at org.easymock.internal.ClassProxyFactory.createProxy(ClassProxyFactory.java:165)
	at org.easymock.internal.MocksControl.createMock(MocksControl.java:107)
	at org.easymock.internal.MocksControl.createMock(MocksControl.java:80)
	at org.easymock.IMocksControl.mock(IMocksControl.java:44)
	at org.easymock.EasyMock.mock(EasyMock.java:70)
	at org.easymock.EasyMock.createMock(EasyMock.java:322)
	at org.easymock.tests.MockClassControlTest.initMock(MockClassControlTest.java:92)
	at org.easymock.tests.MockClassControlTest.testHashCode(MockClassControlTest.java:130)
	at org.easymock.tests.MockClassControlTest.testHashCode_WithOverload(MockClassControlTest.java:122)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:43)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:82)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:73)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:71)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitForkedStarter.main(JUnitForkedStarter.java:29)
Caused by: net.sf.cglib.core.CodeGenerationException: java.lang.reflect.InaccessibleObjectException-->Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @5a61f5df
	at net.sf.cglib.core.ReflectUtils.defineClass(ReflectUtils.java:464)
	at net.sf.cglib.core.AbstractClassGenerator.generate(AbstractClassGenerator.java:339)
	at net.sf.cglib.core.AbstractClassGenerator$ClassLoaderData$3.apply(AbstractClassGenerator.java:96)
	at net.sf.cglib.core.AbstractClassGenerator$ClassLoaderData$3.apply(AbstractClassGenerator.java:94)
	at net.sf.cglib.core.internal.LoadingCache$2.call(LoadingCache.java:54)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at net.sf.cglib.core.internal.LoadingCache.createEntry(LoadingCache.java:61)
	at net.sf.cglib.core.internal.LoadingCache.get(LoadingCache.java:34)
	at net.sf.cglib.core.AbstractClassGenerator$ClassLoaderData.get(AbstractClassGenerator.java:119)
	at net.sf.cglib.core.AbstractClassGenerator.create(AbstractClassGenerator.java:294)
	at net.sf.cglib.core.KeyFactory$Generator.create(KeyFactory.java:221)
	at net.sf.cglib.core.KeyFactory.create(KeyFactory.java:174)
	at net.sf.cglib.core.KeyFactory.create(KeyFactory.java:153)
	at net.sf.cglib.proxy.Enhancer.<clinit>(Enhancer.java:73)
	... 57 more
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @5a61f5df
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
	at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:199)
	at java.base/java.lang.reflect.Method.setAccessible(Method.java:193)
	at net.sf.cglib.core.ReflectUtils$1.run(ReflectUtils.java:61)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:569)
	at net.sf.cglib.core.ReflectUtils.<clinit>(ReflectUtils.java:52)
	at net.sf.cglib.core.KeyFactory$Generator.generateClass(KeyFactory.java:243)
	at net.sf.cglib.core.DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
	at net.sf.cglib.core.AbstractClassGenerator.generate(AbstractClassGenerator.java:332)
	... 69 more

**Comment by henri-tremblay** [comment]
Yes. Right now, you need to open that package in command line. I will give more details when I have a few minutes.

**Comment by sda399** [comment]
> Yes. Right now, you need to open that package in command line. I will give more details when I have a few minutes.

more and more people are migrating to 17, is there any documentation for easymock ?

**Comment by henri-tremblay** [comment]
Not yet. But any PR will be accepted :-)

**Comment by anthraxx** [comment]
Any change you may find some time to investigate the different java 17 problems? Would be nice to see support here

**Comment by henri-tremblay** [comment]
The main reason is that cglib is not compatible with JDK 17.
I have a start of migration to ByteBuddy. I can push the code to a branch for people to help out.
Will do tonight.

**Comment by henri-tremblay** [comment]
I have pushed the branch if someone wants to have a look. It's here https://github.com/easymock/easymock/tree/bytebuddy

**Comment by RangerRick** [comment]
> The main reason is that cglib is not compatible with JDK 17. I have a start of migration to ByteBuddy. I can push the code to a branch for people to help out. Will do tonight.

Hey @henri-tremblay, I just gave this a shot in a portion of our (opennms) codebase, and it appears to work just fine!  I had to add a bunch of `--add-open` bits to the surefire argLine, but it worked.

for reference: `--add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.math=ALL-UNNAMED --add-opens java.base/java.util=ALL-UNNAMED --add-opens java.base/java.util.concurrent=ALL-UNNAMED --add-opens java.base/java.net=ALL-UNNAMED --add-opens java.base/java.text=ALL-UNNAMED`

**Comment by henri-tremblay** [comment]
That's good news. Let me try. However, I'm not sure all this should be needed. Thanks for testing.

**Comment by john9x** [comment]
Hi @henri-tremblay !

I checked bytebuddy version on our projects and it's not work :(
NPE on `replay` call.
```
java.lang.RuntimeException: On mock #2 (zero indexed): Cannot invoke "org.easymock.internal.ObjectMethodsFilter.getDelegate()" because "handler" is null

	at org.easymock.EasyMock.getRuntimeException(EasyMock.java:2223)
	at org.easymock.EasyMock.replay(EasyMock.java:1954)
```
The exceptions on JDK 11 and JDK 17 are same.

It seems that mocking of the interfaces works fine but mocking of the classes not.

**Comment by john9x** [comment]
Hi @henri-tremblay !

I take a look to `org.easymock.internal.ClassProxyFactory#createProxy` and I don't find where you fill `$callback` field with handler. So this is why is not working.
With [my fix](https://github.com/john9x/easymock/commit/5161428e61f8760024b745936d01041c38d1f406) most of all our tests works now.
But next problem is package-private classes, for example:
RuleEvent - is package-private class
```
java.lang.IllegalAccessError: failed to access class com.navixy.server.events.RuleEvent from class com.navixy.server.events.EventChecker$ByteBuddy$CRXtFuXL (com.navixy.server.events.RuleEvent is in unnamed module of loader 'app'; com.navixy.server.events.EventChecker$ByteBuddy$CRXtFuXL is in unnamed module of loader org.easymock.bytebuddy.dynamic.loading.ByteArrayClassLoader @3f5d37e7)

	at com.navixy.server.events.EventChecker$ByteBuddy$CRXtFuXL.<clinit>(Unknown Source)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1042)
	at java.base/jdk.internal.reflect.UnsafeFieldAccessorFactory.newFieldAccessor(UnsafeFieldAccessorFactory.java:43)
	at java.base/jdk.internal.reflect.ReflectionFactory.newFieldAccessor(ReflectionFactory.java:186)
	at java.base/java.lang.reflect.Field.acquireFieldAccessor(Field.java:1105)
	at java.base/java.lang.reflect.Field.getFieldAccessor(Field.java:1086)
	at java.base/java.lang.reflect.Field.set(Field.java:780)
	at org.easymock.bytebuddy.implementation.LoadedTypeInitializer$ForStaticField.onLoad(LoadedTypeInitializer.java:170)
	at org.easymock.bytebuddy.implementation.LoadedTypeInitializer$Compound.onLoad(LoadedTypeInitializer.java:233)
	at org.easymock.bytebuddy.dynamic.TypeResolutionStrategy$Passive.initialize(TypeResolutionStrategy.java:102)
	at org.easymock.bytebuddy.dynamic.DynamicType$Default$Unloaded.load(DynamicType.java:6159)
	at org.easymock.bytebuddy.dynamic.DynamicType$Default$Unloaded.load(DynamicType.java:6147)
	at org.easymock.internal.ClassProxyFactory.createProxy(ClassProxyFactory.java:191)
	at org.easymock.internal.MocksControl.createMock(MocksControl.java:107)
	at org.easymock.internal.MocksControl.createMock(MocksControl.java:80)
	at org.easymock.IMocksControl.mock(IMocksControl.java:44)
	at org.easymock.EasyMock.mock(EasyMock.java:70)
	at org.easymock.EasyMock.createMock(EasyMock.java:322)
```

**Comment by john9x** [comment]
>next problem is package-private classes

one possible solution is to use Unsafe injection

```
.load(toMock.getClassLoader(), new ClassLoadingStrategy.ForUnsafeInjection())
```
This approach resolves problem with package-private classes for me.

Next problems :)
1. Broken partial mocking of abstract classes
```
mock = createMockBuilder(SomeAbstractClass.class).createMock();
expect(mock.someAbstractMethod());  
```
fails with `java.lang.AbstractMethodError: Receiver class SomeAbstractClass$ByteBuddy$inCvNzCx does not define or inherit an implementation of the resolved method 'abstract java.util.List someAbstractMethod()' of abstract class SomeAbstractClass.`
2.
```
ScheduledThreadPoolExecutor executor = createMock(ScheduledThreadPoolExecutor.class);
```
fails with 
```
java.lang.IllegalStateException: java.lang.NoClassDefFoundError: net/sf/cglib/proxy/Factory

	at org.easymock.bytebuddy.dynamic.loading.ClassInjector$UsingUnsafe$Dispatcher$Enabled.defineClass(ClassInjector.java:2023)
	at org.easymock.bytebuddy.dynamic.loading.ClassInjector$UsingUnsafe.injectRaw(ClassInjector.java:1798)
	at org.easymock.bytebuddy.dynamic.loading.ClassInjector$AbstractBase.inject(ClassInjector.java:114)
	at org.easymock.bytebuddy.dynamic.loading.ClassLoadingStrategy$ForUnsafeInjection.load(ClassLoadingStrategy.java:571)
	at org.easymock.bytebuddy.dynamic.TypeResolutionStrategy$Passive.initialize(TypeResolutionStrategy.java:100)
	at org.easymock.bytebuddy.dynamic.DynamicType$Default$Unloaded.load(DynamicType.java:6159)
	at org.easymock.internal.ClassProxyFactory.createProxy(ClassProxyFactory.java:185)
	at org.easymock.internal.MocksControl.createMock(MocksControl.java:107)
	at org.easymock.internal.MocksControl.createMock(MocksControl.java:80)
	at org.easymock.IMocksControl.mock(IMocksControl.java:44)
	at org.easymock.EasyMock.mock(EasyMock.java:70)
	at org.easymock.EasyMock.createMock(EasyMock.java:322)
```
I suppose this is because of `new ByteBuddy().  .implement(Factory.class)`

**Comment by john9x** [comment]
> java.lang.IllegalStateException: java.lang.NoClassDefFoundError: net/sf/cglib/proxy/Factory

[fix](https://github.com/john9x/easymock/commit/db2c373a6e1d64337a6ef8653caed4c7d18f22b5)

>Broken partial mocking of abstract classes
[fix](https://github.com/john9x/easymock/commit/2820396684d02f2585fb382aed1d9e36d1da5c24)

Now all our tests (~7000) passes.

**Comment by henri-tremblay** [comment]
That's awesome! I will go through all of it this week.

**Comment by john9x** [comment]
>one possible solution is to use Unsafe injection

But as far as I understand, using Unsafe is not recommended way. Mockito uses [Lookup + Reflection](https://github.com/mockito/mockito/blob/main/src/main/java/org/mockito/internal/creation/bytebuddy/SubclassInjectionLoader.java#L29) with some kind of magic.

**Comment by henri-tremblay** [comment]
@john9x Was playing with your branch. But tests are not passing for me.
I have
```
[ERROR] Failures: 
[ERROR]   MockedExceptionTest.testExplicitFillInStackTrace:60 expected same:<EasyMock for class java.lang.RuntimeException> was not:<java.lang.RuntimeException>
[ERROR]   MockedExceptionTest.testMockedException:38 expected same:<EasyMock for class java.lang.RuntimeException> was not:<java.lang.NullPointerException: Cannot throw exception because the return value of "java.lang.Throwable.fillInStackTrace()" is null>
[ERROR]   SerializationTest.testClass:74 expected:<a> but was:<null>
[ERROR] Errors: 
[ERROR]   CglibTest.test:47->createMock:62 ExceptionInInitializer
[ERROR]   DefaultClassInstantiatorTest.badSerializable:204 » InaccessibleObject Unable t...
[ERROR]   DefaultClassInstantiatorTest.serializable:198->checkInstantiation:214 » InaccessibleObject
[ERROR]   ClassExtensionHelperTest.testGetControl_EnhancedButNotAMock:61 NoClassDefFound
[ERROR]   ClassExtensionHelperTest.testGetControl_NotAMock:86 » Runtime java.lang.NoSuch...
[ERROR]   EasyMockSupportTest.mockType:332 NoClassDefFound Could not initialize class ne...

```

**Comment by henri-tremblay** [comment]
Ok... Cglib and DefaultClassInstantiator will fail forever. That's fine. Now the others :-)

**Comment by zhangsummer** [comment]
自动回复：
你的邮件我已经收到，现在不在线我会尽早查收的...祝生活愉快/

**Comment by john9x** [comment]
Hi @henri-tremblay 
[fix for mocked exceptions](https://github.com/john9x/easymock/commit/4cfb0a398cd98002122b929a0023333441f5a9cb)

**Comment by john9x** [comment]
[fix cglib related tests](https://github.com/john9x/easymock/commit/154c35594475c86c437e6a7ff69c5e8f8126f699)

I don't know how to fix
```
DefaultClassInstantiatorTest.badSerializable
DefaultClassInstantiatorTest.serializable
SerializationTest.testClass
```

**Comment by fbacchella** [comment]
I met this bug in my own use of easy mock and got a strange work around. With the latest commit, this build fails:

    JAVA_HOME=$(/usr/libexec/java_home -v 18) mvn clean package -Dtest=ValueAxisMrtgTest

with:

    ValueAxisMrtgTest.testBasicEmptyRrd:98->AxisTester.prepareGraph:88 » NoClassDefFound Could not initialize class org.easymock.internal.ClassProxyFactory$2
    ...

but this succeeed:

    JAVA_HOME=$(/usr/libexec/java_home -v 18) mvn clean package -Dtest=ValueAxisMrtgTest -Psonar

The main difference with the profile sonar, it’s that it adds:

    [INFO] --- jacoco-maven-plugin:0.8.8:prepare-agent (prepare-agent) @ rrd4j ---
    [INFO] argLine set to -javaagent:.../.m2/repository/org/jacoco/org.jacoco.agent/0.8.8/org.jacoco.agent-0.8.8-runtime.jar=destfile=.../rrd4j/target/jacoco.exec

I hope this help in debugging.

**Comment by tbalzer** [comment]
Thanks @john9x  and @henri-tremblay for your work. I've tried myself to fix the last few testcases as we would really like to see EasyMock work with Java 17, but I don't see an obvious solution here and I'm not sure they need to be fixed tbh.

This is the point I'm currently at:
* DefaultClassInstatiator might be obsolete in favor of ObjenesisClassInstatiator
* The (presumed) contract of Mock objects being serializable/deserializable might be broken, but I'm not sure if that's a must-have. I would think that upgrading from cglib to bytebuddy would warrant a major version change anyway.

Overall I think the best way might be to remove DefaultClassInstatiator and the SerializationTest and then release the current state as a 5.0-rc1 if nobody has an idea of how to fix the remaining test cases. But I might be missing something.

**Comment by anthraxx** [comment]
@henri-tremblay Any way you could maybe have a final look here, pretty please :cat: I agree with @tbalzer summary here, I think we should just move forward with the new major version and remove those bits.

Thank you very much for your time on easymock

**Comment by henri-tremblay** [comment]
I had some time to play and I'm close to something now. It's all in the bytebuddy branch.

**Comment by anthraxx** [comment]
@henri-tremblay This is amazing news which makes me happy and gives a bit of confidence. I literally was thinking about dropping you a mail today or tomorrow to get infos, as I would need to find some solution to use java17 and it would make me quite sad if that would be without easymock :cat: 
Thank you again for your time and development, this is appreciated a lot :partying_face:

**Comment by henri-tremblay** [comment]
Fixed with #300
